### PR TITLE
[0266] New Vehicle data GearStatus

### LINKED ISSUE #1256: [SDL 0266] New vehicle data GearStatus
Proposal: [New vehicle data GearStatus](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0266-New-vehicle-data-GearStatus.md)

> This proposal is to add new vehicle data `GearStatus` to know drive mode and actual gear the transmission has selected.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/881

Steering Committee Decision:

>The Steering Committee voted to accept this proposal with the following revisions: use `userSelectedGear` in place of `selectedGear` and use the descriptions  for `GearStatus` included in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/881#issuecomment-573838656).
> The proposal .md file was updated to reflect these revisions on 1/21/2020.

### PR DESCRIPTION
**Pull request by kostyaBoss**
Fixes #1256  

This PR is **[ready]** for review.

### Risk
This PR makes **[minor]** API changes.

### Testing Plan
Covered by existing and added unit tests

#### Unit Tests
New unit tests covering introduced structures were added

### Summary
Applied [0266 New Vehicle data GearStatus](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0266-New-vehicle-data-GearStatus.md) changes

#### Enhancements
    - Extend PRNDL enum
    - Add TransmissionType
    - Extend VehicleDataType
    - Implement changes in proper requests responses related to new structs

### CLA
- [ x ] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by kostyaBoss** [comment]
@mrapitis , @atiwari9 please review the changes

**Comment by santhanamk** [comment]
@kostyaBoss I reviewed your PR.  I suggest you make the below changes.

**Source code changes:**

_VehicleDataType.java:_
 - indicate that VEHICLEDATA_PRNDL is deprecated

_GearStatus.java:_
 - since all parameters are not mandatory you don't need to construct the object.
    So you can delete this code - 

    `public GearStatus(
            @NonNull PRNDL userSelectedGear,
            @NonNull PRNDL actualGear,
            @NonNull TransmissionType transmissionType
    ) {
        this();
        setUserSelectedGear(userSelectedGear);
        setActualGear(actualGear);
        setTransmissionType(transmissionType);
    }`

_SubscribeVehicleData:_
 - comment should be "...please see GearStatus" instead of Gear above `getPrndl`

_GetVehicleData:_
 - comment should be "...please see GearStatus" instead of Gear above `getPrndl`

_SubscribeVehicleDataResponse:_
 - prndl needs to be deprecated

_GetVehicleDataResponse:_
 - prndl needs to be deprecated

_OnVehicleData:_
 - prndl needs to be deprecated

_TransmissionType:_
 - change the comment     `/**`
     `* Convert String to PRNDL`
     `* @param value String`
     `* @return PRNDL`
     `*/`

     to

     `/**`
     `* Convert String to TransmissionType`
     `* @param value String`
     `* @return TransmissionType`
     `*/`

**Unit Test changes:**

_Test.java:_
 - missing GearStatus RPC

_GearStatusTests.java:_
 - unit tests for GearStatus need to be added.

_TransmissionTypeTests.java:_
 - I would change `TransmissionType enumSport = TransmissionType.valueForString(example);` to `TransmissionType continuouslyVariable = TransmissionType.valueForString(example);`
 - I would change `TransmissionType continuouslyVariable = TransmissionType.valueForString(example);` to `TransmissionType infinitelyVariable = TransmissionType.valueForString(example);`
 - I would change assert statements from:
        `assertNotNull("PARK returned null", manual);`
        `assertNotNull("REVERSE returned null", automatic);`
        `assertNotNull("NEUTRAL returned null", semiAutomatic);`
        `assertNotNull("DRIVE returned null", dualClutch);`
        `assertNotNull("SPORT returned null", enumSport);`
        `assertNotNull("LOWGEAR returned null", continuouslyVariable);`
        `assertNotNull("FIRST returned null", electricVariable);`
        `assertNotNull("FIRST returned null", directDrive);`

        to:
        assertNotNull("MANUAL returned null", manual);
        assertNotNull("AUTOMATIC returned null", automatic);
        assertNotNull("SEMI_AUTOMATIC returned null", semiAutomatic);
        assertNotNull("DUAL_CLUTCH returned null", dualClutch);
        assertNotNull("CONTINUOUSLY_VARIABLE returned null", enumSport);
        assertNotNull("INFINITELY_VARIABLE returned null", continuouslyVariable);
        assertNotNull("ELECTRIC_VARIABLE returned null", electricVariable);
        assertNotNull("DIRECT_DRIVE returned null", directDrive);

_UnsubscribeVehicleDataResponseTest.java:_
 - change `JSONObject gearStatus = JsonUtils.readJsonObjectFromJsonObject(parameters, SubscribeVehicleDataResponse.KEY_GEAR_STATUS);` to
`JSONObject gearStatus = JsonUtils.readJsonObjectFromJsonObject(parameters, UnsubscribeVehicleDataResponse.KEY_GEAR_STATUS);`

_GetVehicleDataResponseTests.java:_
 - `else if (key.equals(OnVehicleData.KEY_GEAR_STATUS)) {` change to `else if 
     (key.equals(GetVehicleDataResponse.KEY_GEAR_STATUS)) {`

**Comment by kostyaBoss** [comment]
@santhanamk updated according to provided suggestions, please review

**Comment by santhanamk** [comment]
@kostyaBoss I reviewed your PR.  It looks good.  I have a few more suggestions:

**Unit test changes:**

_TransmissionTypeTests.java:_
 - I would rename `infinitely_variable` to `infinitelyVariable` on line 27
 - On line 38 it should be assertNotNull("INFINITELY_VARIABLE returned null", infinitelyVariable);

**Comment by kostyaBoss** [comment]
@santhanamk Hi, thanks for feedback. I've updated the PR, please, review the changes

**Comment by santhanamk** [comment]
@kostyaBoss code looks good.  Thanks.

**Comment by kostyaBoss** [comment]
@joeygrover , @theresalech Hi! Please review changes in this PR

**Comment by theresalech** [comment]
@kostyaBoss the description for this PR says `This PR is [not ready] for review.`  Can you please update if this is ready?  Thanks!

**Comment by santhanamk** [comment]
@theresalech this PR is ready for Livio review.

**Comment by kostyaBoss** [comment]
@RHenigan Hi! I've made corresponding changes, please, check the latest version. Please, let me know if there anything else I should change

**Comment by RHenigan** [comment]
Hello @kostyaBoss, Please focus on this PR #1395
Once this PR is approved other PRs can be updated to match

**Comment by kostyaBoss** [comment]
@RHenigan Hi! Regarding PR #1395, the @bilal-alsharifi has approved it. The problem with "unclear difs" comes from different line [comment](https://github.com/smartdevicelink/sdl_java_suite/pull/1395#issuecomment-669982922). I and @vladmu noticed that issue comes from the wrong line endings, files are containing different types of endings and we can't fix it. There were pushed commits to reduce the number of changes and make difs as clear as possible. There were also commits to align this PR with latest version of the dev branch

cc @bilal-alsharifi , @vladmu

**Comment by RHenigan** [comment]
Hello @kostyaBoss and @vladmu, thank you for making these changes. After approving the WindowStatus PR (#1396) there are now merge conflicts with this PR. Please resolve the merge conflicts and I can review this PR again.

It is likely that as our team reviews other open PRs we will likely run into merge conflicts again. If you only want to do one at a time and have our team review that one so that you’re not updating a bunch at once, that is fine

**Comment by kostyaBoss** [comment]
@RHenigan let's make it one by one. I've updated PR and resolved conflicts, please check

**Comment by RHenigan** [comment]
Hello @kostyaBoss and @vladmu, thank you for resolving the merge conflicts but upon closer inspection there are still large changes in this PR that should not be included please fix these files to minimize the size of the diff.
`GetVehicleData.java`
`GetVehicleDataResponse.java`
`OnVehicleData.java`
`SubscribeVehicleData.java`
`SubscribeVehicleDataResponse.java`
` UnsubscribeVehicleData.java`
`UnsubscribeVehicleDataResponse.java`
`VehicleDataType.java`

**Comment by vladmu** [comment]
@RHenigan 
> there are still large changes in this PR

Fixed. Please review.

**Comment by kostyaBoss** [comment]
@RHenigan applied changes

File: base/src/main/java/com/smartdevicelink/proxy/rpc/OnVehicleData.java [hunk 1]
```diff
@@ -409,12 +423,34 @@ public void setVin(String vin) {
...
+    @Deprecated
     public void setPrndl(PRNDL prndl) {
         setParameters(KEY_PRNDL, prndl);
     }
+
+    /**
+     * Gets the prndl.
+     *
+     * @return PRNDL See PRNDL. This parameter since 7.0.0 is deprecated and it is now covered in `gearStatus`
+     * @since SmartDeviceLink 7.0.0
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/OnVehicleData.java:442]: @kostyaBoss can you also apply these changes?

File: base/src/main/java/com/smartdevicelink/proxy/rpc/SubscribeVehicleData.java [hunk 2]
```diff
@@ -525,23 +532,44 @@ public Boolean getExternalTemperature() {
...
+     */
+    public Boolean getGearStatus() {
+        return getBoolean(KEY_GEAR_STATUS);
+    }
+
+    /**
+     * Sets the prndl.
+     *
+     * @param prndl See PRNDL. This parameter is deprecated since 7.0.0 and it is now covered in `gearStatus`
+     * @since SmartDeviceLink 7.0.0
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/SubscribeVehicleData.java:559]: @kostyaBoss can you also apply these changes?

File: base/src/main/java/com/smartdevicelink/proxy/rpc/SubscribeVehicleData.java [hunk 3]
```diff
@@ -525,23 +532,44 @@ public Boolean getExternalTemperature() {
...
-	 * subscribed.
-	 * 
-	 * @return Boolean -a Boolean value. If true, means the Currently selected gear data
-	 *         has been subscribed.
-	 */
+    /**
+     * Gets the prndl.
+     *
+     * @return Boolean See PRNDL. This parameter is deprecated since 7.0.0 and it is now covered in `gearStatus`
+     * @since SmartDeviceLink 7.0.0
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/SubscribeVehicleData.java:570]: @kostyaBoss can you also apply these changes?

File: base/src/main/java/com/smartdevicelink/proxy/rpc/SubscribeVehicleDataResponse.java [hunk 4]
```diff
@@ -226,16 +227,22 @@ public VehicleDataResult getExternalTemperature() {
         return (VehicleDataResult) getObject(VehicleDataResult.class, KEY_EXTERNAL_TEMPERATURE);
     }
     /**
-     * Gets currently selected gear data
-     * @param prndl a VehicleDataResult related to the PRNDL status (automatic transmission gear)
+     * Sets the prndl.
+     *
+     * @param prndl See PRNDL. This parameter is deprecated and it is now covered in `gearStatus`
+     * @since SmartDeviceLink 7.0.0
```
[the remark this answers is not in the document]
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/proxy/rpc/SubscribeVehicleDataResponse.java:233]: @kostyaBoss can you also apply these changes?

**Comment by RHenigan** [comment]
Hello @kostyaBoss There were 4 that were missed, can you apply those changes as well?
`SubscribeVehicleDataResponse.java`
`SubscribeVehicleData.java`
`OnVehicleData.java`

**Comment by vladmu** [comment]
@kostyaBoss @RHenigan applied missing suggestions. Please review.

**Comment by RHenigan** [comment]
Hello @kostyaBoss and @vladmu Thank you for making the changes.

Please tag us on the next PR you would like us to focus on.

**Pull request merged**